src/layers/normalise.jl

```
@@ -28,26 +30,30 @@ automatically managed using the [`Dropout`](@ref) layer instead of the
 
 The [`Dropout`](@ref) layer is what you should use in most scenarios.
 """
-function dropout(x, p; dims=:, active::Bool=true)
+function dropout(rng::AbstractRNG, x, p; dims = :, active::Bool = true)
   active || return x
-  y = dropout_mask(x, p, dims=dims)
+  y = dropout_mask(rng, x, p, dims=dims)
   return x .* y
 end
 
-@adjoint function dropout(x, p; dims=:, active::Bool=true)
+function dropout(x, p; dims = :, active::Bool = true)
+  dropout(GLOBAL_RNG, x, p, dims = dims, active = active)
+end
+
+@adjoint function dropout(rng, x, p; dims = :, active::Bool = true)
   active || return x, Δ -> (Δ, nothing)
-  y = dropout_mask(x, p, dims=dims)
-  return x .* y, Δ -> (Δ .* y, nothing)
+  y = dropout_mask(rng, x, p, dims = dims)
+  return x .* y, Δ -> (nothing, Δ .* y, nothing)
 end
 
-function dropout_mask(x, p; dims=:)
-  y = rand!(similar(x, _dropout_shape(x, dims)))
+function dropout_mask(rng::AbstractRNG, x, p; dims=:)
+  y = rand!(rng, similar(x, _dropout_shape(x, dims)))
```

I don't think this dispatches to the GPU intrinsic when `rng` is specified. @maleadt?

---

We only override `default_rng`, I haven't looked into getting GLOBAL_RNG to work (maybe it does).

---

We can switch it for `default_rng` too of course

---

But you have to call it on the GPU. You can't call `default_rng` in the Dropout ctor and forward that RNG object to the GPU (unless Dropout objects are created in device code?).

---

I guess it depends on who we give the responsibility to provide a valid rng to (the user or flux) - a proverbial `gpu(rng)`

---

But is this code actually being executed directly on the GPU (i.e. in a kernel)? If not, device overrides and `default_rng` doesn't matter here, and you do actually need to provide a GPU RNG for use on the host, e.g., `CUDA.RNG()` or `CUDA.CURAND.RNG()`. "write your kernel once and run anywhere" doesn't really work if it isn't a kernel.

---

Sorry, but I'm confused. If I call `rand!(x)` and `x isa CuArray`, then I'll dispatch to the GPU override. If I call `rand!(rng, x)` and `rng` is one of the many RNGs available in Julia, then this obviously cannot run on the GPU, but I would consider that a user error. But, what should the default value of `rng` be such that when the user doesn't customize it, it will call `Random.rand!` for `Array`s and `CUDA.rand!` for `CuArray`s? Or do we need to set the default conditionally based on the type of `x`?

---

Alternatively to setting the default conditionally, we can make `gpu(::Dropout)` change `GLOBAL_RNG` to `CUDA.RNG()` (maybe this is what you meant @DhairyaLGandhi).

---

> But, what should the default value of `rng` be such that when the user doesn't customize it, it will call `Random.rand!` for `Array`s and `CUDA.rand!` for `CuArray`s? Or do we need to set the default conditionally based on the type of `x`?

There currently isn't an API to figure out a good RNG for an array type, so this would need some additional methods. `default_rng` isn't such an API because it doesn't take arguments; I had presumed this code is executing in a GPU kernel where we use contextual dispatch (think Cassette) to replace the method and return an appropriate RNG. But that's not relevant here, since this is host code using GPU functionality instead of code directly executing on the GPU.

---

I have added a dispatch for CUDA. If we are executing a forward pass, I had assumed the whole call graph can get compiled at once doing the method swaps as necessary, but maybe that's not the case - as Tim mentioned it explicitly needs to be part of a CUDA kernel call.

---

I am not sure this is the approach I would go for. It would be weird if I created `Dropout` with `MersenneTwister` and the program silently used `CUDA.CURAND.default_rng()`. Instead, I would use the approach you initially suggested so that `gpu(m)` swaps `GLOBAL_RNG`/`Random.default_rng()` to `CUDA.CURAND.default_rng()` (and errors when `rng` is set to some other variant that cannot be run on the GPU).

My only concern is that we [`fmap(CUDA.cu, ...)`](https://github.com/FluxML/Flux.jl/blob/335286adf118b61ad6fffa5937bd9358477a00c9/src/functor.jl#L72) and defining `CUDA.cu(GLOBAL_RNG)` seems like a big no-no. Maybe we can do the following:
```julia
_isleaf(::AbstractArray{<:Number}) = true
_isleaf(::AbstractArray{T}) where T = isbitstype(T)
_isleaf(::AbstractRNG) = true
_isleaf(x) = false

to_gpu(x) = CUDA.cu(x)
to_gpu(::GLOBAL_RNG) = CUDA.CURAND.default_rng()
to_gpu(rng::AbstractRNG) = error("Cannot move $rng to GPU (only `GLOBAL_RNG` is supported).")

gpu(x) = use_cuda[] ? fmap(to_gpu, x; exclude = _isleaf) : x
```

---

See https://github.com/FluxML/Flux.jl/pull/1618#discussion_r652780581

This approach also ignores a gou rng, so I'm not in favour. Either way, I don't think `to_gpu` is a good api, and the most common case of `Dropout |> gpu` with a cuda compatible rng wouldn't be handled this way

---

I don't really follow the relevance of the linked comment. AFAIK there are no GPU compatible RNGs. There are simply the RNGs provided by CUDA.jl/GPUArrays.jl. `to_gpu` is similar to `trainable` — it allows us to define how to map leaf in the structure to the GPU without defining `CUDA.cu` on types that it shouldn't be defined on. It's out of necessity, but if you can think of an alternative, go for it (I'm not tied to `to_gpu` at all).

---

I was referring to the exactly the ones provided by CUDA/CURAND. Sorry if that wasn't clear.

---

So I don't understand your comment then. This provides a mapping from `GLOBAL_RNG` to `CUDA.CURAND.default_rng()`. And when a user does `Dropout |> gpu` with any other CPU RNG, it throws an error instead of silently ignoring it.

---

I think the most sensible place to error is when the user tries `m |> gpu` on a model that contains `Dropout` with CPU RNGs that have no GPU equivalent. Either way, I don't care how we do it, but as long as we don't silently use a CUDA RNG when `x isa CuArray` and throw an error instead, then it's fine by me.

---

Right - it will error anyway, its more clear when we can say that CUDA needs a way to dispatch to a CUDA RNG. I agree on that.